### ISSUE TITLE
Restricting Container Access from Everyone

### ISSUE THREAD

**Issue by milanchotai** [moby/moby#35984]
I've searched through the issues list, online, etc. and can't find a concrete answer to this, so wanted to ask: How do folks using Docker to distribute their applications on hosts they don't control (ie: a developer's laptop) restrict access to their IP (ie: images / containers)? 

I assumed originally a container could be configured to not allow anyone (including `root`) to access a container that's been "locked-down" (ie: no `exec`, etc.), but the best answer I can find is "You can't, you can only lock-down the host."  

I'm seeking confirmation that it is impossible, or if it is possible, a thread indicating how to do it. Otherwise, for others in my position, I want to know what is the wisest strategy - do we use a VM with Docker installed within it, and lock that down instead?

Hoping I haven't duplicated what I assumed would be a common challenge, but after checking the issue list for Moby, I didn't find anything.

Also, if there is any information about ways to protect Docker images pulled-down to the host machine from being unarchived / decompiled, etc., I would be grateful.

**Comment by AkihiroSuda** [comment]
You may want to use authz plugin.
e.g. https://github.com/twistlock/authz

**Comment by thaJeztah** [comment]
While an `authz` plugin would restrict access to the docker API, someone with `root` access on the host would still have access to anything.

The TL;DR if you don't control the host, you can't control what's on it.

Docker isn't magic, and neither are containers; containers are designed to prevent processes _inside_ the container from breaking _out_ to the host, but not to prevent `root` from _entering_ the container, or access its filesystem.

Images itself are (in transit) `tar.gz` archives (so someone with pull-access on the image, can extract the image), and once downloaded, are "files on disk"; so `root` can access those files. This is basically no different than publishing an `.rpm` or `.deb` package for your software; giving people access to those packages, gives them access to the files inside.

Here's an earlier discussion https://github.com/moby/moby/issues/23200

Please keep in mind that the GitHub issue tracker is not intended as a general support forum,
but for [reporting bugs and feature requests](https://raw.githubusercontent.com/moby/moby/master/.github/ISSUE_TEMPLATE.md). For other type of questions, consider using one of;

- the Docker Support Forums - https://forums.docker.com
- the [Docker community Slack channel](https://blog.docker.com/2016/11/introducing-docker-community-directory-docker-community-slack/) (register here: [http://dockr.ly/community](http://dockr.ly/community))
- [StackOverflow](https://stackoverflow.com)

I'm closing this issue because this is not a bug, but feel free to continue the conversation 👍

**Comment by milanchotai** [comment]
Thanks for clarify @thaJeztah - from what I'd read online this seemed the case, so just wanted to verify before I dug further.

Apologies to post as an issue and thanks for clarifying where to direct future inquiries! :)
